### ISSUE TITLE
[HL/HL:BS/HL:OF/TFC/DOD/CS/RICOCHET] gl_overbright

### ISSUE THREAD

**Issue by Snuffkun** [ValveSoftware/halflife#1891]
Is it really so difficult to make working this in OpenGL? I know many people are missing the old graphics. Without overbright, the levels look wrong, the skyboxes look too bright compared to the levels.

**Comment by Enokilis** [comment]
If the engine finds the OpenGL extension string of GL_ARB_multitexture, it will make use of hardware multitexturing capabilities to draw lightmaps and detail textures.
Unfortunately, this breaks the functionality of gl_overbright, and the only way to restore it is to manually disable the OpenGL extension at the driver level. The lacking functionality of this feature is often incorrectly attributed to a total removal in the Steam version of the game, but the reason for its presence in older versions of the game is that the engine instead checked for an OpenGL extension string named GL_SGIS_multitexture, and because this string is obsolete and not reported in modern graphics drivers, the engine assumes that multitexturing is not supported.
A simple option to toggle multitexturing independently of its driver availability in the game would eliminate this problem.

**Comment by Matthaiks** [comment]
Workaround: https://www.vogons.org/viewtopic.php?f=8&t=63930

Not perfect, but it works.

**Comment by rtxa** [comment]
That seems very cool, also it looks like it fixes issues of low fps with dynamic lights like flashlight, etc.

Quote from that link: https://www.vogons.org/viewtopic.php?f=8&t=63930
> As a sort of experiment, using GLIntercept (https://github.com/dtrebilco/glintercept), and using the command gl_dump in console to get a list of OpenGL extensions that the game detected, I attempted to remove a number of extensions until I came across one: "GL_ARB_multitexture". By removing this specific extension from being detected and used by the game, not only did it seemingly fix the performance around the lights, it also fixes gl_overbright so that it actually works on maps using that feature. Kind of surprised since overbrights don't normally function in the GL renderer.

**Comment by Enokilis** [comment]
> I did a little research to figure out why overbright is forced off when multitexture support is available.

I'm pretty sure the lack of overbright functionality with multitexturing comes down to OpenGL limitations.
The basic ARB multitexture extension does not have all of the blending modes that exist in the base OpenGL specification, namely `glBlendFunc( GL_DST_COLOR, GL_SRC_COLOR )`. As far as I know, the only way to achieve the same effect in one rendering pass is to use the texture combiner extension, as I did here. 
![gl_overbright](https://user-images.githubusercontent.com/29983795/193400776-4f2c1c0a-f205-4118-a5cf-061d7c0f9e5c.jpg)

**Comment by SamVanheer** [comment]
Makes sense. I suppose it would be easier to implement in a shader then since you can control the blending more precisely with any number of textures.

**Comment by Enokilis** [comment]
> Makes sense. I suppose it would be easier to implement in a shader then since you can control the blending more precisely with any number of textures.

It's also worth noting that gl_overbright never worked well to begin with. In the software renderer, all lit surfaces are affected, but when activated in the accelerated renderer, vertex-lit objects remain capped at 100% brightness, due to how this effect relies on blending two passes together. This creates an undesirable disparity.

There are also certain quirks in how the engine processes lightmaps with consequences for the visual fidelity. When gl_overbright is disabled, the engine tries to compensate by boosting the brightness of the lightmaps at the cost of dynamic range. This behavior was not present in the early release versions of the game. The VHLT map compiler takes this modern behavior into consideration at the detriment of restored overbright functionality.

**Comment by megakarlach** [comment]
As of November 17/18, 2023 update (25th Anniversary), 
https://github.com/ValveSoftware/halflife/issues/3424

It's seem like the overbright issue was not entirely fixed, only the chromatic aberration issue was fixed.
